transport/pool.go

```diff
@@ -0,0 +1,145 @@
+package transport
+
+import (
+	"errors"
+	"runtime"
+	"scylla-go-driver/frame"
+	. "scylla-go-driver/frame/request"
+	. "scylla-go-driver/frame/response"
+	"time"
+)
+
+type ConnPool struct {
+	conns []*Conn
+}
+
+type PoolRefiller struct {
+	address    string
+	connConfig ConnConfig
+	connPool   ConnPool
+
+	activeShards uint16
+	nrShards     uint16
+
+	connOptions      frame.ScyllaSupported
+	errRcv           chan uint16
+	errorSinceRefill bool
+	currentDelay     time.Duration
+}
+
+const (
+	minFillBackoff        = 50 * time.Millisecond
+	maxFillBackoff        = 10 * time.Second
+	fillBackoffMultiplier = 2
+)
+
+func (p *PoolRefiller) loop() { // nolint:unused // This will be used.
+	runtime.LockOSThread()
+
+	p.currentDelay = minFillBackoff
+	ticker := time.NewTicker(p.currentDelay)
+	for {
+		select {
+		case <-ticker.C:
+			p.fill()
+		case nr := <-p.errRcv:
+			p.removeConnection(nr)
+		}
+
+		if p.needsFilling() {
+			if p.errorSinceRefill {
+				p.onFillError()
+			} else {
+				p.currentDelay = minFillBackoff
+			}
+
+			ticker.Reset(p.currentDelay)
+		} else {
+			ticker.Stop()
+		}
+	}
+}
+
+func (p *PoolRefiller) onFillError() {
+	if maxFillBackoff > p.currentDelay*fillBackoffMultiplier {
+		p.currentDelay *= fillBackoffMultiplier
+	} else {
+		p.currentDelay = maxFillBackoff
+	}
+}
+
+func (p *PoolRefiller) needsFilling() bool { // nolint:unused // This will be used.
+	return p.activeShards < p.nrShards
+}
+
+func (p *PoolRefiller) removeConnection(nr uint16) { // nolint:unused // This will be used.
+	err := p.connPool.conns[nr].conn.Close()
+	p.errorSinceRefill = p.errorSinceRefill || (err != nil)
+	p.connPool.conns[nr] = nil
+	p.activeShards--
+}
+
+func (p *PoolRefiller) fill() {
```

Comment:
I find it pretty long. 
How about moving the functionality of this whole initiation to some other function. 
Maybe somewhere in InitNodeConnPool.